File: scrapers/weworkremotely.py
"""
We Work Remotely RSS feed scraper.
"""
import logging
import feedparser
from bs4 import BeautifulSoup
from datetime import datetime
from .base import Job
from config import WWR_FEEDS, SEARCH_KEYWORDS

logger = logging.getLogger(__name__)


def scrape_weworkremotely(keywords: list[str] | None = None) -> list[Job]:
    if keywords is None:
        keywords = SEARCH_KEYWORDS

    kw_lower = [k.lower() for k in keywords]
    results: list[Job] = []
    seen: set[str] = set()

    for feed_url in WWR_FEEDS:
        try:
            feed = feedparser.parse(feed_url)
        except Exception as e:
            logger.warning(f"[WWR] Feed {feed_url} error: {e}")
            continue

        for entry in feed.entries:
            title_raw = entry.get("title", "")
            description_html = entry.get("summary", "") or ""
            link = entry.get("link", "")

            if link in seen:
                continue

            # Title format: "Company: Position"
            company, title = _split_title(title_raw)

            description = BeautifulSoup(description_html, "html.parser").get_text(
                separator=" ", strip=True
            )[:1000]

            searchable = f"{title} {company} {description}".lower()
            if not any(kw in searchable for kw in kw_lower):
                continue

            posted = _parse_struct_time(getattr(entry, "published_parsed", None))

            seen.add(link)
            results.append(Job(
                title=title,
                company=company,
                url=link,
                source="WeWorkRemotely",
                job_type="remote",
                location="Remote",
                description=description,
                posted_date=posted,
            ))

    logger.info(f"[WeWorkRemotely] Found {len(results)} relevant jobs")
    return results


def _split_title(raw: str) -> tuple[str, str]:
    """'Company: Position' → (company, position)."""
    if ": " in raw:
        parts = raw.split(": ", 1)
        return parts[0].strip(), parts[1].strip()
    if " - " in raw:
        parts = raw.split(" - ", 1)
        return parts[0].strip(), parts[1].strip()
    return "", raw.strip()


def _parse_struct_time(st) -> datetime | None:
    if st is None:
        return None
    try:
        return datetime(*st[:6])
    except Exception:
        return None

File: scrapers/weworkremotely.js
import RSSParser from "rss-parser";
import { load } from "cheerio";
import { Job } from "./base.js";
import { WWR_FEEDS, SEARCH_KEYWORDS } from "../config.js";

const parser = new RSSParser({ timeout: 20000 });

export async function scrapeWeWorkRemotely(keywords = SEARCH_KEYWORDS) {
  const kwLower = keywords.map((k) => k.toLowerCase());
  const results = [];
  const seen = new Set();

  for (const feedUrl of WWR_FEEDS) {
    try {
      const feed = await parser.parseURL(feedUrl);
      for (const entry of feed.items || []) {
        const link = entry.link || "";
        if (seen.has(link)) continue;

        const [company, title] = splitTitle(entry.title || "");
        const descHtml = entry.content || entry.summary || entry["content:encoded"] || "";
        const description = load(descHtml).text().trim().slice(0, 1000);

        const searchable = `${title} ${company} ${description}`.toLowerCase();
        if (!kwLower.some((kw) => searchable.includes(kw))) continue;

        seen.add(link);
        results.push(
          new Job({
            title,
            company,
            url: link,
            source: "WeWorkRemotely",
            jobType: "remote",
            location: "Remote",
            description,
            postedDate: entry.pubDate ? new Date(entry.pubDate) : null,
          })
        );
      }
    } catch (err) {
      console.warn(`[WWR] Feed error ${feedUrl}: ${err.message}`);
    }
  }

  console.log(`[WeWorkRemotely] ${results.length} relevant jobs`);
  return results;
}

function splitTitle(raw) {
  if (raw.includes(": ")) {
    const [co, ...rest] = raw.split(": ");
    return [co.trim(), rest.join(": ").trim()];
  }
  if (raw.includes(" - ")) {
    const idx = raw.indexOf(" - ");
    return [raw.slice(0, idx).trim(), raw.slice(idx + 3).trim()];
  }
  return ["", raw.trim()];
}
